[WiP] Unbundling plugins

tfennelly opened the pull request:
[As discussed on the dev mailing list](https://groups.google.com/forum/?hl=en#!topic/jenkinsci-dev/kRobm-cxFw8).

**TODOs**:
- [x] Support/utility code to help determine the type of install/upgrade is required. See `StartupUtil` class (and test class).
- [x] Wire in simple plugin selection wizard that just lets you install a set of predefined plugins (selection feature comes later). Relies on the code in `StartupUtil` to decide how to act.
- [x] Perform plugin upgrade. Relies on the code in `StartupUtil` to decide how to act.
- [x] Recognise downgrades.
- [x] Fix test harness tests.
- [x] "REST" API for connection status (JSON).
- [x] "REST" API for installing plugins via JSON post + ability to track install of a set of plugins via a correlationId.
- [x] "REST" API for getting plugin install status (JSON).
- [x] Version 1.0 of the JS lib plugins.
- [x] Add "temp" install wizard so people can test the PR build from CI.
- [x] Deprecate pinning [ala dev list](https://groups.google.com/d/msg/jenkinsci-dev/kRobm-cxFw8/6V66uhibAwAJ).
- [x] Localise the wizard.
- [x] Add real install wizard.
- [x] Better handling of page reload while install in progress.
- [x] Add jslint.
- [x] ~~Proxy support.~~ deferred, messaging included in offline mode.
- [x] Show plugin dependencies.
- [x] Error handling - failed plugin installs - Jenkins crashes (OOMEs etc)
- [x] Offline support. User should be guided towards a KB article or other.
- [x] Test with a range of supported browsers (including IE of course).

tfennelly committed:
Added StartupUtil

tfennelly committed:
Removed unused utility functions

tfennelly committed:
Added hasStartedSinceUnbundlingEpoc

And set some @since javadoc

tfennelly committed:
Added getDetachedPlugins from ClassicPluginStrategy

tfennelly committed:
Javadoc updates

tfennelly commented:
@reviewbybees

KostyaSha commented on core/src/main/java/jenkins/install/StartupType.java:
```diff
@@ -0,0 +1,45 @@
+/**
+ * Jenkins Start type.
+ *
+ * @author <a href="mailto:tom.fennelly@gmail.com">tom.fennelly@gmail.com</a>
```
Why are you using personal email if you are using CloudBees copyright for license?

tfennelly replied:
It's the email I go by on the dev list and other places Jenkins related.

oleg-nenashev commented:
Seems you're updating the <code> ClassicPluginStrategy</code> only. Any plans to bring such functionality into other strategies? Otherwise the advantage of the new API methods seems to be quite limited

daniel-beck commented:
> Seems you're updating the `ClassicPluginStrategy` only. Any plans to bring such functionality into other strategies? Otherwise the advantage of the new API methods seems to be quite limited

`ClassicPluginStrategy` is the only implementation I (and the extension indexer) know. Do you know others?

oleg-nenashev commented:
> ClassicPluginStrategy is the only implementation I (and the extension indexer) know. Do you know others?

No. But it would worth to add the new methods to the Extension point's API

daniel-beck commented on core/src/main/java/hudson/ClassicPluginStrategy.java:
```diff
@@ -280,9 +282,36 @@ protected ClassLoader createClassLoader(List<File> paths, ClassLoader parent, At
      * Information about plugins that were originally in the core.
      */
-    private static final class DetachedPlugin {
+    public static final class DetachedPlugin {
```
`@Restricted(NoExternalUse.class)`

jglick replied:
Not if it is going to be returned from `getDetachedPlugins`, in which case you need a `@since` on this too.

tfennelly commented:
> Seems you're updating the `ClassicPluginStrategy` only. Any plans to bring such functionality into other strategies? Otherwise the advantage of the new API methods seems to be quite limited

The code changed in `ClassicPluginStrategy` is static code relating to detached plugins. Tbh, I think it doesn't really belong in there, but didn't move it because thought it would end up being a point of silly controversy.

daniel-beck commented on core/src/main/java/jenkins/install/StartupUtil.java:
```diff
@@ -0,0 +1,153 @@
+     * This state information is required in order to determine whether or not the Jenkins instance
+     * is just restarting, or is being upgraded from an earlier version.
+     */
+    public static void saveLastExecVersion() {
```
Nobody seems to ever call this (other than your test)?

tfennelly replied:
This is a WiP. It'll be used in further changes.

daniel-beck replied:
Right. See my comment in the PR thread suggesting a placeholder comment for that, which would allow seeing this functionality in action.

daniel-beck commented on core/src/main/java/jenkins/install/StartupUtil.java:
```diff
@@ -0,0 +1,153 @@
+     * <p>
+     * This check can be used to determine what kind of upgrade needs to be performed i.e.
+     * <ul>
+     *     <li>Pre unbundling epoc version upgrade: In this case, all plugins ever detached from Jenkins need to be installed.</li>
```
It is not clear to me why there is a need for this distinction. Since Jenkins started out by extracting everything that's bundled all the time, this seems unnecessary.

tfennelly replied:
@daniel-beck So are you saying that the detached plugins should always be installed?

This would seem wrong wrong to me. Seems like the only time they all should be installed should be when the last running instance was an instance that also installed them all i.e. an instance of a release from before bundles were "unbundled".

daniel-beck replied:
@tfennelly No, I'm saying that _today_ Jenkins extracts all bundled plugins every time it starts. Therefore I'd expect it to be able to switch to "only extract what's new" without making this distinction.

tfennelly replied:
@daniel-beck ok thanks. I'm obviously missing something or not asking the right questions.

Here are the 2 upgrade scenarios as I see them...

__Scenario 1__:
Upgrading from an old'ish Jenkins version (that has bundled plugins) e.g. 1.609. Which of the previously bundle plugins (that are no longer bundled) should be auto-installed? I'd have thought you'd need to auto-install them all because you don't know what they've been using.

__Scenario 2__:
Upgrading from a Jenkins version that was released after we unbundled all plugins e.g. upgrading from 1.700 to 1.710 (i.e. some time in the future). So in this situation, they've already gone though at least one upgrade sequence. Which of the previously bundle plugins (that are no longer bundled) should be auto-installed? I'd have thought this is a different upgrade and only plugins that have been detached since 1.700 would need to be auto-install. You can't just install them all again coz you might end up reinstalling something that they uninstalled after the last upgrade (to 1.700).

Where am I wrong there?

tfennelly replied:
Sorry @daniel-beck, I see what you mean now wrt scenario 1. The bit I was missing was that I was assuming (for some reason I don't know) that all bundled plugins were removed on shutdown. Because of that I was thinking they would need to be auto-installed on upgrade where they are no longer bundled. Sorry for the confusion.

tfennelly commented:
> @oleg-nenashev No. But it would worth to add the new methods to the Extension point's API

Read the above comment. The changed code is not related to the extension point API imo (or even the class in which it is located).

daniel-beck commented on core/src/main/java/jenkins/install/StartupUtil.java:
```diff
@@ -0,0 +1,153 @@
+     * </ul>
+     *
+     * @return {@code true} if "this" instance of Jenkins has started a Jenkins instance that was released since
+     * the unbundling epoc, otherwise {@code false}.
```
It's *epoch*. (I know what you wrote, but I'd rather not see an extra commit for this.)

daniel-beck commented:
> the new methods to the Extension point's API

@oleg-nenashev Detached plugins have always been specific to `ClassicPluginStrategy`.

daniel-beck commented:
So while you mention this is WIP and you don't yet have the logic to install things, I'd really like to see at least a log statement on startup that says "This upgraded from X" or "Nothing was updated" or so as placeholder for that.

amuniz commented:
Everything looks good to me so far Tom.

tfennelly committed:
Popup fake/temporary plugin wizard for a "NEW" jenkins

tfennelly commented:
Build is currently broken because it is depending on a temp/hack install wizard (https://github.com/tfennelly/temp-install-wizard).

tfennelly committed:
Extract non web install function in PluginManager (from doInstall)

tfennelly committed:
temp-wizard plugin now uses handlebars

tfennelly committed:
Getting UPGRADE to work

tfennelly committed:
DOWNGRADE support

We don't actually do anything other than update the persisted version file.

tfennelly committed:
Added XML utilities to get values from an XML file.

tfennelly committed:
Changed to get the last executed version directly from config.xml

Waiting for Jenkins to load is too late in the startup sequence and changing how it's done is a PITA because other things start breaking. This is lower risk all around.

tfennelly committed:
Get recommended plugin installs done

tfennelly committed:
Bundling unbundled plugins in test-harness

tfennelly committed:
Getting test-harness tests to pass after unbundling

tfennelly committed:
Get the install wizard JS path from web.xml

tfennelly committed:
UpdateCenter connection status

tfennelly committed:
UpdateCenter connection status - more tests

tfennelly committed:
UpdateCenter install plugins "REST" endpoint

tfennelly committed:
UpdateCenter install plugins "REST" endpoint

tfennelly committed:
UpdateCenter plugin install status "REST" endpoint

tfennelly committed:
Bundling detached compatibility plugins in /WEB-INF/detached-plugins

tfennelly committed:
Refix test-harness tests to use the rebundled detached-plugins

tfennelly committed:
Removed existing node stuff from war build

tfennelly committed:
Added the temporary wizard - real one soon

tfennelly committed:
version 1.0 of the JS Lib plugins

tfennelly committed:
jQuery 1.1 plugin (un minified jQuery)

stephenc commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -423,6 +434,160 @@ public void run(Reactor reactor) throws Exception {
+     *     <li>Are dependencies of one of the above e.g. script-security is not one of the detached plugins but it must be loaded if matrix-project is loaded.</li>
+     * </ul>
+     */
+    protected void loadDetachedPlugins() {
```
This seems to be code duplication. any reason you didn't use https://github.com/jenkinsci/optional-plugin-helper-module to load them?

stephenc replied:
It is a module precisely so that it can be pulled into core via the dependency mechanism

tfennelly replied:
@stephenc I thought it was doing something different and it didn't seem to handle plugin dependencies. There are no docs with it so reading the code is the only way of making sense of it - maybe I didn't read it closely enough.

tfennelly replied:
Ok, I had another look. So it seems like I'd need to implement a `PluginSource` extension point, right? I could do that and it wold work I'm sure but it looks like I'd still need to implement most of the same code anyway (resolving dependencies etc) so it seems like it would actually be complicating it more.

stephenc replied:
Nope it should resolve the dependencies for you. Your source just needs to provide a list of all the plugins it wants and the rest will be resolved for you

tfennelly replied:
Ok, I'll give it a try. If it makes things cleaner then I'm happy to use it.

tfennelly replied:
@stephenc  I added the `PluginSource` and moved the relevant code into it. I was able to test it (using `-Dlicense.disableCheck` - it has no license on it) but seems like I need to add more code (to resolve missing deps perhaps?). At this stage, I already have more code than what it's replacing + it seems to be making it more messy/unreadable so I'm skipping it for now.

stephenc replied:
That sounds like you are certainly doing it wrong. Should be nothing to do with any license at all. I'll take a peek tomorrow

tfennelly replied:
The license is a lesser thing and I'm sure it can be worked out ... I just added the dependency and got the error when trying to build it.

The main thing is ... pulling in an external dependency to replace 20 lines of code with 20+ lines of code doesn't real seem like a win to me.

jglick replied:
Agreed that you should be using the module.

tfennelly replied:
If it improves things I've no problem using it. I don't agree with using it for the sake of using it just coz it exists, especially if it results in more code and added dependencies, which is what seems to be the case to me.

I'll have a look again.

jglick commented on core/src/main/java/hudson/ClassicPluginStrategy.java:
```diff
@@ -300,6 +330,47 @@ private DetachedPlugin(String shortName, String splitWhen, String requireVersion
+        /**
+         * Get the short name of the plugin.
+         * @return The short name of the plugin.
+         * @since FIXME
```
`@since` unnecessary on individual members of a class which was newly exposed.

jglick commented on core/src/main/java/hudson/ClassicPluginStrategy.java:
```diff
@@ -300,6 +330,47 @@ private DetachedPlugin(String shortName, String splitWhen, String requireVersion
+            for (ClassicPluginStrategy.DetachedPlugin detachedPlugin : detachedPlugins) {
+                pluginNames.add(detachedPlugin.getShortName());
+            }
+            return pluginNames;
```
Seems rather dubious to be adding a new API method for what is essentially nothing more than

```java
ClassicPluginStrategy.getDetachedPlugins().stream().map(ClassicPluginStrategy.DetachedPlugin::getShortName).collect(Collectors.toList())
```

tfennelly replied:
Why do you think that's better?

jglick replied:
Because this method is just adding conceptual weight without really adding anything you could not do trivially without it. It is not even called from any new code in this PR that I can see! And APIs without demonstrated uses are always a bad idea.

tfennelly replied:
Well, "trivial" for one person can be obscure to others. I'm not clever enough to be able to look at your alternative and immediately twig "oh yes ... that's just extracting a list of plugin names", so we'll have to agree to disagree on that one :smile:

Anyway ... your last point is what's most relevant to me because yes, that (and the method before, which I also added) do not seem to be used anymore so we can remove them. I must have been using them at an earlier stage in this PR.

jglick commented on core/src/main/java/hudson/ClassicPluginStrategy.java:
```diff
@@ -320,7 +391,7 @@ private void fix(Attributes atts, List<PluginWrapper.Dependency> optionalDepende
     }
 
-    private static final List<DetachedPlugin> DETACHED_LIST = Arrays.asList(
+    private static final List<DetachedPlugin> DETACHED_LIST = Collections.unmodifiableList(Arrays.asList(
```
Or `ImmutableList.of`.

jglick commented on core/src/main/java/hudson/Functions.java:
```diff
@@ -467,6 +468,15 @@ public static boolean isWindows() {
+     * @return The URL path to the Install Wizard JavaScript.
+     * @since FIXME
+     */
+    public static String getInstallWizardJS() {
```
Probably we can avoid dumping new methods like this into the public API. Either use `@Restricted`, or just use a Groovy view which can call the underlying function anyway.

jglick commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -364,13 +375,13 @@ protected void reactOnCycle(PluginWrapper q, List<PluginWrapper> cycle)
-            builder,new TaskGraphBuilder() {{
-            requires(PLUGINS_LISTED).attains(PLUGINS_PREPARED).add("Loading plugins",new Executable() {
+                builder, new TaskGraphBuilder() {{
+            requires(PLUGINS_LISTED).attains(PLUGINS_PREPARED).add("Loading plugins", new Executable() {
```
Gratuitous.

jglick commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -364,13 +375,13 @@ protected void reactOnCycle(PluginWrapper q, List<PluginWrapper> cycle)
                  */
                 public void run(Reactor session) throws Exception {
-                    Jenkins.getInstance().lookup.set(PluginInstanceStore.class,new PluginInstanceStore());
+                    Jenkins.getInstance().lookup.set(PluginInstanceStore.class, new PluginInstanceStore());
```
Gratuitous.

jglick commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -778,45 +945,115 @@ public HttpResponse doUpdateSources(StaplerRequest req) throws IOException {
+     * @since FIXME
+     */
+    @RequirePOST
+    public JSONObjectResponse doInstallPlugins(StaplerRequest req) throws IOException {
```
`@Restricted(NoExternalUse.class)`

tfennelly replied:
Why? This is called by the install wizard, which I assume would be considered to be "external". Other Jenkins distros will also want to call this and they'd def be "external". Why do `@Restricted` for this method and not do it for the other `install` method endpoint (called by the form post action in the PM)?

jglick commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -778,45 +945,115 @@ public HttpResponse doUpdateSources(StaplerRequest req) throws IOException {
-                            }
-                            p = plugin;
+                    UpdateSite.Plugin plugin = getPlugin(pluginName, siteName);
+                    // TODO: Someone that understands what the following logic is about, please add a comment.
```
I guess it means we do not know which update site it is supposed to be loaded from, when >1 are configured.

jglick commented on core/src/main/java/hudson/model/UpdateCenter.java:
```diff
@@ -230,6 +273,74 @@ public InstallationJob getJob(Plugin plugin) {
+     * update site.
+     * 
+     * @return The current connection status.
+     * @since FIXME
```
Better use `@Restricted(NoExternalUse.class)` (or even `DoNotUse`) for things intended to be REST calls only.

tfennelly replied:
Curious to know why you'd treat REST-like endpoints differently to the other endpoint methods that get invoked via http calls?

tfennelly replied:
Done anyway.

jglick replied:
> why you'd treat REST-like endpoints differently to the other endpoint methods

I would not.

tfennelly replied:
Ok, but it happens all over the place, no? `doFillTypeItems`, `doFillNameItems` etc etc (a few random ones I picked from workflow code) and most/many are not marked as `@Restricted` yet they are all web methods, no?.

But again ... I agree with you. I think all of these should be marked `@Restricted`. I'll be trying to do it from now on !

jglick commented on core/src/main/java/jenkins/install/StartupUtil.java:
```diff
@@ -0,0 +1,159 @@
+            try {
+                return FileUtils.readFileToString(lastExecVersionFile);
+            } catch (IOException e) {
+                throw new IllegalStateException("Unexpected Error. Unable to read " + lastExecVersionFile.getAbsolutePath(), e);
```
Recover more gracefully than this.

tfennelly replied:
What would you do? It seems like things are badly screwed up if that file exists and is not readable, no?

jglick replied:
Perhaps; we do not really know. Are you positive you want to halt startup over this?

tfennelly replied:
Jesse, I'm never positive about anything ... I'm full of self doubt :wink: 

Ok ... I could change it to keep going and treat it like a restart (i.e. do nothing wrt install of plugins etc) if you think that makes more sense.

jglick commented on core/src/main/java/jenkins/install/StartupUtil.java:
```diff
@@ -0,0 +1,159 @@
+
+        // Neither the top level config or the lastExecVersionFile have a version
+        // stored in them, which means it's a new install.
+        if (lastRunVersion.compareTo(NEW_INSTALL_VERSION) == 0) {
```
Works, but seems like it would be more intuitive for `getLastExecVersion()` to simply return null if it cannot find a number.

tfennelly replied:
Yeah maybe. I was following on from the fact that Jenkins already had the concept of a "1.0" version for a jenkins instance that never had its global config saved.

jglick commented on core/src/main/java/jenkins/install/StartupUtil.java:
```diff
@@ -0,0 +1,159 @@
+    private static String getCurrentExecVersion() {
+        if (Jenkins.VERSION.equals(Jenkins.UNCOMPUTED_VERSION)) {
+            // This should never happen!! Only adding this check in case someone moves the call to this method to the wrong place.
+            throw new IllegalStateException("Unexpected call to StartupUtil.getCurrentExecVersion(). Jenkins.VERSION has not been initialized. Call computeVersion() first.");
```
Should rather make all of `StartupUtil` be `@Restricted(NoExternalUse.class)` so we know for sure how it is called.

tfennelly replied:
Sure, will do.

jglick commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -872,6 +882,33 @@ protected void doRun() throws Exception {
+     * @return The URL path to the Install Wizard JavaScript.
+     * @since FIXME
+     */
+    public String getInstallWizardJS() {
```
`@Restricted(NoExternalUse.class)`. (Or even get this out of `Jenkins` entirely, unless you are actually using it for Stapler binding.)

tfennelly replied:
Why?

jglick commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -872,6 +882,33 @@ protected void doRun() throws Exception {
+     * Save the last execute version file.
+     * @since FIXME
+     */
+    public void doSaveLastExecVersion() {
```
`@Restricted(NoExternalUse.class)`

tfennelly replied:
Why?

tfennelly replied:
Actually ... I think I might be able to get rid of this altogether, so don't bother justifying your comment. Thanks.

jglick commented on core/src/main/java/jenkins/util/JSONObjectResponse.java:
```diff
@@ -0,0 +1,120 @@
+ *
+ * @author <a href="mailto:tom.fennelly@gmail.com">tom.fennelly@gmail.com</a>
+ */
+public class JSONObjectResponse implements HttpResponse {
```
Should rather be in `HttpResponses`, ultimately the one in Stapler.

tfennelly replied:
Okidoki.

jglick commented on test/src/main/java/org/jvnet/hudson/test/HudsonTestCase.java:
```diff
@@ -501,7 +506,19 @@ protected Hudson newHudson() throws Exception {
+            pluginManager = new LocalPluginManager(home) {
+                @Override
+                protected Collection<String> loadBundledPlugins() {
+                    // Overriding so we can force loading of the detached plugins for testing
```
Could this not just be done with a special `pluginManager` in the particular test suite that is doing that?

tfennelly replied:
It probably could but that wouldn't cover tests outside Jenkins core that are depending on this. I think there was only one test using it inside Jenkins core.

jglick replied:
OK…are you _expecting_ some plugin test to care about this?

jglick replied:
I guess I should ask—is this something you are doing in order that a newly added test can verify detached plugin loading; or something you are doing because otherwise existing tests (esp. those with `@LocalData`) will start breaking?

tfennelly replied:
I did it because one of the tests (that turns on `useLocalPluginManager`) failed and I'm saying there may be other tests outside Jenkins core that may also be using this (but I don't know).

Tbh, this doesn't seem like something worth spending time debating.

tfennelly replied:
I didn't do it randomly for fun :smile:

jglick replied:
> one of the existing tests […] failed

All I needed to know, thanks.

jglick commented on test/src/main/java/org/jvnet/hudson/test/JenkinsRule.java:
```diff
@@ -444,6 +452,10 @@ public void after() throws Exception {
             clients.clear();
 
         } finally {
+            // Make sure the unit test flag gets reset in case a test turned
```
BTW _reset_ in the context of a `boolean` field is normally understood to mean _set to `false`_, making this comment a little confusing.

tfennelly replied:
> reset in the context of a boolean field is normally understood to mean set to false

Really? I'd usually think of it to mean "set back to its default/initial value" (which in this case is `true`), but would interpret it in the context of the code it's connected to.

I think it's perfectly fine having it here but will remove and do it `@After` in `StartupUtilTest`.

jglick commented on test/src/main/java/org/jvnet/hudson/test/JenkinsRule.java:
```diff
@@ -911,6 +923,117 @@ public DumbSlave createSlave(Label l, EnvVars env) throws Exception {
+        if (url.startsWith("/jenkins/")) {
+            url = url.substring("/jenkins/".length());
+        } else if (url.startsWith("/")) {
+            url = url.substring(1);
```
This is wrong. The `url` should always be given relative to the context path, as `goTo` and many other methods expect.

tfennelly replied:
Sorry, what's wrong? This check should not be here?

jglick replied:
Correct, the caller is responsible for specifying a `url` relative to the context path.

tfennelly replied:
Ok, but then it's not "wrong" to have it there (imo) unless we introduce context paths that start with "jenkins", which would be sick.

I'll remove the `startsWith("/jenkins/")` fixup, but I'm leaving the other.

jglick commented on test/src/main/java/org/jvnet/hudson/test/JenkinsRule.java:
```diff
@@ -911,6 +923,117 @@ public DumbSlave createSlave(Label l, EnvVars env) throws Exception {
+        if (url.startsWith("/jenkins/")) {
+            url = url.substring("/jenkins/".length());
+        } else if (url.startsWith("/")) {
+            url = url.substring(1);
```
Also wrong.

jglick commented on test/src/test/java/hudson/model/UpdateCenterPluginInstallTest.java:
```diff
@@ -0,0 +1,92 @@
+    }
+        
+    @Test
+    @RandomlyFails("Will fail if it can't connect to the UC")
```
Prefer direct use of `Assume`/`AssumptionViolatedException`.

jglick commented on test/src/test/java/jenkins/install/StartupUtilTest.java:
```diff
@@ -0,0 +1,110 @@
+        // so remove it so we can fake the tests.
+        StartupUtil.getLastExecVersionFile().delete();
+        // Disable the unit test flag.
+        Main.isUnitTest = false;
```
Better for this to be fixed in `@After` rather than in `JenkinsRule` I think.

tfennelly replied:
Okidoki

tfennelly replied:
Done

jglick commented on war/.gitignore:
```diff
@@ -1,2 +1,10 @@
-work
+work
```
Why does this file even exist? Entries should be moved to top level.

KostyaSha replied:
I thought that war module existed for building war. Shouldn't all UI trash go to separate ui maven module?

tfennelly replied:
@jglick I assume it was added so as not to pollute the top level `.gitignore` with entries that are only relevant to this module.

tfennelly replied:
@KostyaSha Feel free to start a dev list thread on the topic. The pattern has been in play for some time now.

KostyaSha commented on war/pom.xml:
```diff
@@ -538,36 +571,67 @@ THE SOFTWARE.
+                            <phase>initialize</phase>
                             <goals>
-                                <goal>exec</goal>
+                                <goal>wget</goal>
```
wget on windows?

jglick replied:
It is not literally running the `wget` command.

KostyaSha replied:
Node download exists in `frontend-maven-plugin` why another one plugin is required?

KostyaSha replied:
Found before and commented.

jglick commented on war/pom.xml:
```diff
@@ -575,5 +639,124 @@ THE SOFTWARE.
+                                <nodeVersion>v${node.version}</nodeVersion>
+                                <npmVersion>${npm.version}</npmVersion>
+                                <!-- Use the pre-download node and npm packages. See download-maven-plugin config above. -->
+                                <downloadRoot>file:${project.build.directory}/frontend/</downloadRoot>
```
Constructing a `file`-protocol URL in this way is unsafe. (Consider a checkout of Jenkins in a directory `\\workserver\share\My 100% kosher directory name #1\`.) Instead use `${project.baseUri}target/frontend/`.

tfennelly replied:
Cool, thanks.

KostyaSha commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -423,6 +434,160 @@ public void run(Reactor reactor) throws Exception {
+        ServletContext context = Jenkins.getInstance().servletContext;
+        Set<String> plugins = Util.fixNull((Set<String>) context.getResourcePaths(fromPath));
+        Set<URL> copiedPlugins = new HashSet<>();
+        Set<URL> dependencies = new HashSet<>();
```
URL in hashset?

tfennelly replied:
No need.

KostyaSha replied:
AFAIK hashCode in URL calls external resources that will lead to long delays.

jglick replied:
FindBugs will indeed warn about it. The problem is very real if the URLs are, for example, `http` protocol, because `URL.equals` does a DNS comparison on hostnames, so just adding to a collection could block on network I/O! Dumb, right? (`URI` uses a more superficial comparison and so is safe.) In this case I think the URLs are always local so it is not an issue.

tfennelly replied:
Wtf ... that's screwed up. I wasn't aware of that. Yes, all of these URLs are classpath resource URLs, so will not be `http`.

@KostyaSha sorry for my earlier reply. I thought you were referring to the fact that `new HashSet<>()` didn't define the generic type.

KostyaSha commented on test/src/test/java/hudson/model/UpdateCenterConnectionStatusTest.java:
```diff
@@ -0,0 +1,161 @@
+    };
+    
+    @Test
+    public void test_states_allok() {        
```
is it some methodology to name test methods in java in C-style?

KostyaSha commented on war/pom.xml:
```diff
@@ -538,36 +571,67 @@ THE SOFTWARE.
+      
+    <!--
+        Manually download node and NPM for the frontend plugin so it gets cached in the local maven repo.
+        Otherwise frontend will be downloading it all the time on the CI servers. The download-maven-plugin 
```
npm modules usually contains different trash and directly downloads from internet that wouldn't be cached.  So if you want really cache everything, this wouldn't help.

KostyaSha commented on war/pom.xml:
```diff
@@ -538,36 +571,67 @@ THE SOFTWARE.
+    -->
+      
+    <!--
+        Manually download node and NPM for the frontend plugin so it gets cached in the local maven repo.
```
on CI server you should also cache npm registry

KostyaSha commented:
:-1: for custom caching mechanism, if you care about cache in CI build, then you should create 2 cache repositories and point maven-frontend-plugin:
```
                                    <nodeDownloadRoot>${nodeDownloadRoot}/</nodeDownloadRoot>
                                    <npmDownloadRoot>${npmRegistry}/npm/-/</npmDownloadRoot>
```
First should be generic proxy, second npm proxy.

tfennelly commented:
@KostyaSha what is you understanding of what is being cached? From your comments, it sounds like you're talking about caching of NPM modules, which is not what's going on in the pom.

Either way ... I don't mind removing that but I'm sure others would be annoyed at the frontend plugin re-downloading the node and npm packages.

tfennelly committed:
Updates based on comments from @jglick

tfennelly committed:
Use `Assume` in `UpdateCenterPluginInstallTest`

tfennelly committed:
Added a few `@Restricted(NoExternalUse.class)` as requested by @jglick

tfennelly committed:
Added i18n RootAction

tfennelly committed:
Added i18n RootAction - fixed test and added more javadoc

tfennelly committed:
Added i18n RootAction - more javadoc

jglick commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -778,45 +944,117 @@ public HttpResponse doUpdateSources(StaplerRequest req) throws IOException {
+     * @since FIXME
+     */
+    @RequirePOST
+    public HttpResponse doInstallPlugins(StaplerRequest req) throws IOException {
```
Should be `@Restricted(DoNotUse.class)`.

> This is called by the install wizard, which I assume would be considered to be "external".

But that is an HTTP “call”, not a Java “call”.

> Why do `@Restricted` for this method and not do it for the other `install` method endpoint

All web methods should be similarly restricted.

tfennelly replied:
Ok, we're getting subtle :smile: (but it is a good point).

The thing is .... I see `doThis` and `doThat` methods all over Jenkins (all web methods) and it was never a pattern until now (or I just haven't been aware of it perhaps). I'm not complaining though coz I do think it makes sense (I'm sure there would be some exceptions of course) to restrict these methods and not have Java based calls to them. The ones that take a `StaplerRequest` (and maybe return a `HttpResponse`) are 99.9% guaranteed to fit this.

jglick commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -872,6 +882,36 @@ protected void doRun() throws Exception {
     /**
+     * Get the Jenkins {@link StartupType startup type}.   
+     * @return The Jenkins {@link StartupType startup type}.
+     * @since FIXME
```
`@since` is pointless (indeed misleading) on a `@Restricted` method. If for some reason it ever needs to be unrestricted in the future, you would add `@since` for _that_ version.

tfennelly replied:
Fair point. Thanks.  I added the `@since` earlier and then never thought to remove it once I added the `@Restricted`.

Commit:
JENKINS-9598 - Initial installation wizard & curated list of plugins

tfennelly committed:
Added @since markers

Commit:
Fix Node.js downloads for different Windows architectures

Commit:
Fix 32-bit Node.js downloads, too.

tfennelly committed:
More @Restricted fixups

tfennelly committed:
Treat as RESTART if unable to read last exec version file.

tfennelly committed:
Removed triming of "/jenkins/" from the front of JenkinsRule.getJSON and .postJSON calls

Keeping @jglick happy ;)

tfennelly commented:
@jglick thanks for the review ... I think I've addressed all your concerns (or at least the latest round of them ;) ).

tfennelly committed:
Deprecate plugin pinning

tfennelly committed:
Fixed detached plugin install bug

Doing an explicit installed Vs bundled version check now.

Commit:
Merge remote-tracking branch 'primary/unbundling-plugins' into unbundling-plugins

tfennelly committed:
Added correlationId info to the install status response

tfennelly committed:
Added correlationId info to the install status response - null check

Commit:
Remove unnecessary 'usage' data, fix 'offline' flash in certain cases,
always use new API for installation status

jglick commented:
> I see `doThis` and `doThat` methods all over Jenkins (all web methods) and it was never a pattern until now

Do not assume that historical precedents are right just by virtue of being older. In many cases these were written before `@Restricted` even existed.

I am tempted to just mark them all `@Restricted` retroactively. Source-incompatible but binary-compatible.

jglick commented on test/src/main/java/org/jvnet/hudson/test/JenkinsRule.java:
```diff
@@ -911,6 +919,113 @@ public DumbSlave createSlave(Label l, EnvVars env) throws Exception {
+        JenkinsRule.WebClient webClient = createWebClient();
+
+        if (url.startsWith("/")) {
+            url = url.substring(1);
```
> it's not "wrong" to have it there (imo)

It is inconsistent with that is expected of `goTo` and similar methods, which specifically forbid an initial slash, so that callers must recognize that they are passing a context-relative path.

jglick commented on core/src/main/java/jenkins/diagnostics/PinningIsBlockingBundledPluginMonitor.java:
```diff
@@ -1,44 +0,0 @@
- * @author Kohsuke Kawaguchi
- */
-@Extension
-public class PinningIsBlockingBundledPluginMonitor extends AdministrativeMonitor {
```
You forgot to delete its associated resource directory.

tfennelly replied:
Done. Thanks.

tfennelly committed:
Removed PinningIsBlockingBundledPluginMonitor resource dir

tfennelly committed:
Make JenkinsRule.getJSON and JenkinsRule.postJSON consistent with goTo methods

Less friendly... more picky, but hey imo

tfennelly committed:
Merge branch 'unbundling-plugins' into kzantow-unbundling-plugins

* unbundling-plugins:
  Make JenkinsRule.getJSON and JenkinsRule.postJSON consistent with goTo methods
  Removed PinningIsBlockingBundledPluginMonitor resource dir
  Added correlationId info to the install status response - null check
  Added correlationId info to the install status response

tfennelly commented:
@jglick 

> Do not assume that historical precedents are right just by virtue of being older. In many cases these were written before @Restricted even existed.

Don't get me wrong ... we are in agreement. I think this is a good thing. I just didn't hear anyone mention it before and I like to know where things come from. Also, the fact that the javadoc on `Restricted` says "Indicates that a particular element is really __*deprecated*__ ..." (emp. __*deprecated*__) made me wonder. Maybe that javadoc should be changed to reflect the broader use of `Restricted`.

> I am tempted to just mark them all `@Restricted` retroactively. Source-incompatible but binary-compatible.

I think it would be good to make it clearer in the code why such methods are restricted by having a restriction value of something like `WebOnly` (or whatever) e.g. `@Restricted(WebOnly.class)`.

In reality though, I think it's near impossible for anyone to invoke many of these methods from Java code because constructing a usable `StaplerRequest` etc is not trivial.

tfennelly commented:
.... although we know that my idea of "trivial" is not the same as yours, so you might think it is trivial :smiley:

tfennelly committed:
gulpfile cleanups

tfennelly committed:
Center buttons

tfennelly committed:
Tweak welcome screen text

Reduce the amount of words

kzantow committed:
Modularize server callbacks; add individual selected plugin progress
indicators during installation, including errors; style tweaks

kzantow committed:
Setup wizard: Add loading panel while connectivity check runs, fix issue
with IE9 caching AJAX responses; better responsiveness for the search &
select top area

tfennelly committed:
Wait until plugins are installed before saving last exec file

tfennelly committed:
Renamed StartupType to InstallState and StartupUtil to InstallUtil

tfennelly committed:
Get page reload working while install in progress

 Still not 100% right ... using the "available" plugin list is not right because it means you get a different list as you are installing (those that are installed drop off the available list).

tfennelly committed:
removed console log

tfennelly committed:
Added "pluginManager/plugins" RESTish endpoint and using it in wizard

Versus using the "availables" endpoint, which was only returning the plugins that were not installed.

Commit:
Gus adding styling touches

tfennelly committed:
Tweak welcome message again

tfennelly committed:
Some light JavaScript refactorings

Attempting to make some parts a bit clearer.

tfennelly committed:
Added some comments

tfennelly committed:
Added some info on plugin selection re it not being the full list

tfennelly committed:
Added some info on plugin selection re it not being the full list - part 2

jglick commented:
> I think it's near impossible for anyone to invoke many of these methods from Java code because constructing a usable `StaplerRequest` etc is not trivial.

It is true that coming up with a mock `StaplerRequest` would be very hard. However many web methods do not need, and do not request, such a parameter.

```java
@RequirePOST
public HttpResponse doWhatever(@QueryParameter String someArgument) {…}
```

jglick commented on test/src/main/java/org/jvnet/hudson/test/JenkinsRule.java:
```diff
@@ -911,6 +919,114 @@ public DumbSlave createSlave(Label l, EnvVars env) throws Exception {
+        assert !path.startsWith("/");
+
+        JenkinsRule.WebClient webClient = createWebClient();
+        Page runsPage = null;
```
I think it is legal to drop the initializer here.

jglick commented:
Patch looks OK AFAICT. Have not tried it myself.

kzantow committed:
Setup wizard: add ability to resume if jenkins is restarted/dies in the
middle of installation; add better error handling; add close button; add
restart when necessary after plugin installation

tfennelly commented on core/src/main/java/jenkins/install/InstallUtil.java:
```diff
@@ -0,0 +1,230 @@
+     */
+	public static void clearInstallStatus() {
+		persistInstallStatus(null);
+	}
```
@kzantow I think it would be good to update `InstallUtilTest` with some unit tests for the new "status" file utilities.

tfennelly committed:
Small tweaks - method name match-ups, additional javadocs etc

kzantow committed:
Modify offline messaging & styles; correct issue closing installer

kzantow committed:
Add more installer tests

kzantow committed:
Add jslint & fix lint issues; fix dependency tooltips

kzantow committed:
Merge remote-tracking branch 'primary/master' into unbundling-plugins

tfennelly committed:
Fix PluginManagerInstalledGUITest after merge of plugin manager dependency changes

tfennelly committed:
JSHint from jenkins-js-builder

And fixed jshint issues

kohsuke committed:
Added AssetManager to serve static assets from classpath

Kinda like AdjunctManager, except that the resource lookup rule
is modified to allow plugins to bring their own assets that can
overlap with the core.

tfennelly committed:
Modify AssetManager to only load from the Jenkins core classpath

tfennelly committed:
Handle 404 in AssetManager

daniel-beck commented:
@tfennelly Could you please resubmit this towards the `2.0` branch?

tfennelly committed:
Load Jenkins JS Modules from 'core-assets'

tfennelly committed:
Merge asset manager stuff

tfennelly committed:
Fixed version

daniel-beck commented:
It looks like we're getting a build time dependency on *node* (and will exclude e.g. FreeBSD as build platform AFAIU). Will need to be documented accordingly on the wiki once this gets merged.

tfennelly committed:
Updated bootstrap dep to catch snapshot dep fix

tfennelly committed:
TestPluginManager.loadBundledPlugins dir exists check

tfennelly committed:
Added a plugin to @Test so as to generate API return data

tfennelly commented:
Merged over to the 2.0 branch.
